Linstor volume plugin

rp- opened the pull request:
### Description

This PR adds a Linstor volume(primary storage) cloudstack driver to cloudstack.
Basic operations as creating/migrating/deleting are working.
Snapshots work with Linstor for most storage pool types, but will fail because cloudstack doesn't seem to support
snapshots of RAW image types(suggestions/ideas are welcome)

It also has a commit to port elastistor jersey-client to version 2.26, so it will not collide with Linstor dependencies (I can put that in a separate PR if needed)

Design doc: https://cwiki.apache.org/confluence/display/CLOUDSTACK/Linstor+storage+plugin

Fixes: #4403 

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

### Feature/Enhancement Scale or Bug Severity

#### Feature/Enhancement Scale

- [X] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### How Has This Been Tested?
4 Node cloudstack cluster (1 Mgmt node, 3 KVM-hosts) with equivalent Linstor setup(1 Controller, 3 Satellites)
Attaching volumes to VM's, migrating volumes between NFS storage and Linstor, creating templates from Linstor volumes

rhtyd commented:
Thanks for the PR @rp- is this ready for review/testing?

rp- commented:
Yes, but I'm also open for design discussions as most of the code was done by looking at other storage plugins and trying to understand(debug) how cloudstack handles most things.
So if there are stupid or unnecessary things in my code please don't be a afraid to point me into the right directions.

rhtyd commented:
@rp- a can you at least raise a documentation PR (https://github.com/apache/cloudstack-documentation) to explain how to use this and some details on how this was tested. It would be great if you can share a design document of this feature (for example, https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=158873968 a new storage plugin).

I can help kick regression tests in the meanwhile.
@blueorangutan package

blueorangutan commented:
@rhtyd a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: :heavy_check_mark: centos7 :heavy_check_mark: centos8 :heavy_check_mark: debian. SL-JID 514

rhtyd commented:
@blueorangutan test

blueorangutan commented:
@rhtyd a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian test result (tid-604)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 50396 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr4994-t604-kvm-centos7.zip
Intermittent failure detected: /marvin/tests/smoke/test_reset_vm_on_reboot.py
Intermittent failure detected: /marvin/tests/smoke/test_resource_accounting.py
Intermittent failure detected: /marvin/tests/smoke/test_router_dhcphosts.py
Intermittent failure detected: /marvin/tests/smoke/test_routers_network_ops.py
Intermittent failure detected: /marvin/tests/smoke/test_vpc_redundant.py
Intermittent failure detected: /marvin/tests/smoke/test_hostha_kvm.py
Smoke tests completed. 82 look OK, 6 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
ContextSuite context=TestResetVmOnReboot>:setup | `Error` | 0.00 | test_reset_vm_on_reboot.py
ContextSuite context=TestRAMCPUResourceAccounting>:setup | `Error` | 0.00 | test_resource_accounting.py
test_router_dhcphosts | `Failure` | 216.62 | test_router_dhcphosts.py
ContextSuite context=TestRouterDHCPHosts>:teardown | `Error` | 226.91 | test_router_dhcphosts.py
test_01_RVR_Network_FW_PF_SSH_default_routes_egress_true | `Failure` | 306.12 | test_routers_network_ops.py
test_03_create_redundant_VPC_1tier_2VMs_2IPs_2PF_ACL_reboot_routers | `Failure` | 423.77 | test_vpc_redundant.py
test_disable_oobm_ha_state_ineligible | `Error` | 1511.56 | test_hostha_kvm.py

sureshanaparti commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/storage/LinstorStoragePool.java:
```diff
@@ -0,0 +1,195 @@
+        String name,
+        QemuImg.PhysicalDiskFormat format,
+        Storage.ProvisioningType provisioningType,
+        long size)
```
@rp- please ensure the code format is in-line with the current formatting, and is consistent.

rp- replied:
Do you have an exact formatting guideline somewhere?
I tried to bring them a bit more in line now

sureshanaparti commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/storage/LinstorStoragePool.java:
```diff
@@ -0,0 +1,195 @@
+    @Override
+    public KVMPhysicalDisk getPhysicalDisk(String volumeUuid)
+    {
+        return this._storageAdaptor.getPhysicalDisk(volumeUuid, this);
```
I think, no need to explicitly mention _this_  for class method / variable unless one exists with same name in that scope.

sureshanaparti commented on plugins/storage/volume/linstor/src/main/java/org/apache/cloudstack/storage/datastore/driver/LinstorPrimaryDataStoreDriverImpl.java:
```diff
@@ -0,0 +1,420 @@
+            VolumeInfo volumeInfo = (VolumeInfo) dataObject;
+
+            long storagePoolId = dataStore.getId();
+            final String rscName = "cs-" + volumeInfo.getPath();
```
Better to define the prefix "_cs-_" as string constant in some common class, and use it across.

sureshanaparti commented on plugins/storage/volume/linstor/src/main/java/org/apache/cloudstack/storage/datastore/driver/LinstorPrimaryDataStoreDriverImpl.java:
```diff
@@ -0,0 +1,420 @@
+                    break;
+                case TEMPLATE:
+                    s_logger.debug("createAsync - creating template");
+                    break;
```
If the driver doesn't create snapshot / template here, better do not handle these cases.

rp- replied:
Hmm I wouldn't completely remove the cases, but I added an errMsg, so the callback gets an error answer,
If I would remove the cases, it would return an InvalidDataobject error, which it isn't

sureshanaparti commented on plugins/storage/volume/linstor/src/main/java/org/apache/cloudstack/storage/datastore/lifecycle/LinstorPrimaryDataStoreLifeCycleImpl.java:
```diff
@@ -0,0 +1,327 @@
+
+    @Override
+    public boolean attachZone(DataStore dataStore, ZoneScope scope, HypervisorType hypervisorType) {
+        List<HostVO> hosts = resourceMgr.listAllUpAndEnabledHostsInOneZoneByHypervisor(HypervisorType.KVM,
```
Hypervisor hard-coded to KVM here, check with the _hypervisorType_ input and if it is supported or not.

sureshanaparti commented on plugins/storage/volume/linstor/src/main/java/org/apache/cloudstack/storage/datastore/lifecycle/LinstorPrimaryDataStoreLifeCycleImpl.java:
```diff
@@ -0,0 +1,327 @@
+    public boolean attachCluster(DataStore dataStore, ClusterScope scope) {
+        PrimaryDataStoreInfo primaryDataStoreInfo = (PrimaryDataStoreInfo) dataStore;
+
+        // check if there is at least one host up in this cluster
```
Check if this datastore supports the hypervisor in the cluster or not.

rp- commented:
> @rp- a can you at least raise a documentation PR (https://github.com/apache/cloudstack-documentation) to explain how to use this and some details on how this was tested. It would be great if you can share a design document of this feature (for example, https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=158873968 a new storage plugin).

@rhtyd 
I created an account at cwiki.apache.org, but I'm not sure how, or if I can create such a design document there yet?

rhtyd commented:
@rp- can you join our mailing list and drop a line to dev@ http://cloudstack.apache.org/mailing-lists.html I think someone will access will grant you permission to add/edit pages (cc @DaanHoogland - do you have management access or do we ask ASF infra?)

@blueorangutan package

blueorangutan commented:
@rhtyd a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: :heavy_check_mark: centos7 :heavy_check_mark: centos8 :heavy_check_mark: debian. SL-JID 523

weizhouapache commented on ui/public/locales/en.json:
```diff
@@ -3060,6 +3061,7 @@
 "message.reset.password.warning.notpasswordenabled": "The template of this instance was created without password enabled",
 "message.reset.password.warning.notstopped": "Your instance must be stopped before attempting to change its current password",
 "message.reset.vpn.connection": "Please confirm that you want to reset VPN connection",
+"message.resourcegroup.description": "Linstor resource group to use for primary storage",
```
can you add "linstor" into the key "message.resourcegroup.description" ?

GutoVeronezi commented:
@rp- I noticed that you added a lot of comments to the code. Most of they only describe what the next line is doing, e.q.: 
```java
//verify if var is true then print it in console
if (var) {
    System.out.println("var is true");
}
```

These types of comments do not improve debugging and goes against a clean code. However, there are comments that explain a bit of the process, but they could be resolved with an extraction of the code to a method with a corresponding name, javadoc and etc.

Also, there are `TODO` comments. These requests should be regularized in issues, not in code.

Comments should be a last resort, not the rule. If we take all the comments and sort then in a logic way, we could write a book :smile: .

GutoVeronezi commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/storage/LinstorStorageAdaptor.java:
```diff
@@ -0,0 +1,566 @@
+        int timeout)
+    {
+        s_logger.info("Linstor: createDiskFromTemplate");
+        // if source is linstor lvm-thin snapshot we could snapshot and restore from, but do plain copy for now
```
@rp- I do not know Linstor and do not know why you choose to use this way, but I am wondering why do a plain copy instead of a snapshot...

GutoVeronezi replied:
@rp- ok, understood.

I know from `qemu` that there is a way to coalesce (merge) or collapse the snapshots ([cheatsheet](https://blog.programster.org/qemu-img-cheatsheet)). Could ZFS have anything like that?

rp- commented:
I added a basic PR for documentation and also a first version of the design document:
https://cwiki.apache.org/confluence/display/CLOUDSTACK/Linstor+storage+plugin
(sorry I mostly copied the stuff from the ScaleIO plugin)

Btw. do you have some storage test guidelines? Like what basic functions should work and how to test them?
I'm not to familiar with all Cloudstack functionalities, so chances are high that I simply missed some functionality tests.

GutoVeronezi commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/storage/LinstorStorageAdaptor.java:
```diff
@@ -0,0 +1,566 @@
+
+            StringJoiner sj = new StringJoiner(System.getProperty("line.separator"));
+            reader.lines().iterator().forEachRemaining(sj::add);
+            result = sj.toString();
```
I did a little search and found a similar usage in `Agent.java` https://github.com/apache/cloudstack/blob/6b9f3fb5015c29bbfaee0b08220174541ce40ab0/agent/src/main/java/com/cloud/agent/Agent.java#L475-L478

If we go deeper we may find another implementations.

I would suggest you to create an agent utility and implement a method to retrieve the `hostname`, then use it instead of this whole method.

sureshanaparti commented:
> I added a basic PR for documentation and also a first version of the design document:
> https://cwiki.apache.org/confluence/display/CLOUDSTACK/Linstor+storage+plugin
> (sorry I mostly copied the stuff from the ScaleIO plugin)
> 
> Btw. do you have some storage test guidelines? Like what basic functions should work and how to test them?
> I'm not to familiar with all Cloudstack functionalities, so chances are high that I simply missed some functionality tests.

@rp- Please check if Linstor storage plugin supports all the functionalities mentioned in the cwiki. Can you add some storage references in cwiki ?

rp- commented:
> @rp- Please check if Linstor storage plugin supports all the functionalities mentioned in the cwiki. Can you add some storage references in cwiki ?

What do you mean with "storage references" ?

sureshanaparti commented:
> > @rp- Please check if Linstor storage plugin supports all the functionalities mentioned in the cwiki. Can you add some storage references in cwiki ?
> 
> What do you mean with "storage references" ?

@rp- I mean some references to know about the Linstor storage. I found one mentioned in cwiki page, and  updated the references section.

rhtyd commented:
@rp- here are some links for you for learning/references:  cc @Philipp-Reisner
https://github.com/shapeblue/hackerbook
https://github.com/shapeblue/mbx

Here are some high-level lifecycle operations a typical storage provider can support: (it also depends if which ones of the following operations are even possible/allowed by the storage control plane)
- Volumes: list (by name, id etc), create, resize, clone, delete, migrate, map/unmap/check-map (to a host), download volume (via secondary storage), get volume/disk statistics/metrics (usage, iops etc)
- Volume Snapshots: list, take snapshot, revert snapshot, delete (note: this is a single volume/disk), download snapshot (via secondary storage) 
- VM Snapshots: list, take snapshot, revert snapshot, delete (note: this may require a consistency group, i.e. a group of volumes/disks that are attached to a disk with or without memory, i.e. disk-only VM snapshot and disk+ram snapshots)
- Template: create template from snapshot, create template from volume/disk
- Storage pool (or Primary storage): list (by name, id etc), get pool statistics/metrics (usage, iops etc), create/add (with label)
- Migration: can we migrate volumes from one storage pool (Linstore resource pool) to another pool (Linstore resource pool or other types of pool), can we migrate VM across hosts/clusters with/without Linstor storage

Most of the above operations already have a high level CloudStack API, generally the storage provider/plugin implements the backend for these operations. In some cases, if an operation is not allowed (for example, raw-disk snapshots on KVM) you may need to add conditional checks or refactor the CloudStack service layers (API, managers, storage sub-system... which largely defines the policy) and hypervisor sub-system (KVM+storage specific storage data motion strategy, LibvirtStoragePoolDef, handling of config drive, VM/disk migration, KVM specific linstore storage pool manager + storage processor + storage adaptor classes, i.e. they implement how storage/hypervisor specific behaviour are executed).

In general you would setup a dev-test env, and while adding new storage pool in CloudStack add a label to that pool (primary storage) and then create disk offerings with the label - this way you can force VM's root&data disks to be created on the pool. These are called storage tags (we've something similar called host tags which can be used to map compute/service offerings with hypervisor hosts). Lastly, you want to write Marvin-based integration tests specific to your storage provider.

Example implementations for references:
https://github.com/apache/cloudstack/pull/4304 (ScaleIO storage provider)
https://github.com/apache/cloudstack/tree/master/plugins/storage/volume (all current storage providers)
https://github.com/apache/cloudstack/tree/master/test/integration/plugins (examples of plugin-specific integration tests)
https://cwiki.apache.org/confluence/display/CLOUDSTACK/Storage+subsystem+2.0 (CloudStack storage framework)

Hope this helps.

rhtyd commented:
Ping @rp- any update for us - is this ready for review/testing, and functionally complete from your side?
@blueorangutan package

blueorangutan commented:
@rhtyd a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

rp- commented:
The force push 11 days ago, was my new version with snapshots working, but before I announce it here, I wanted to give it a final test in my development installation. But suddenly as I wanted to start the rebased version, cloudstack stopped working on my end.
Last week I was on vacation till today and I just retried a fresh rebased version, but cloudstack still doesn't want to come online:
[management-server.log](https://github.com/apache/cloudstack/files/6731497/management-server.log)

No error in the log and on the web interface port I get `HTTP ERROR 503 Service Unavailable`, maybe you know whats wrong here?

slavkap commented:
Hi @rp-, I have faced such a problem when there are merged PRs with new tables/added or renamed columns between your commits. You could check what's new here `engine/schema/src/main/resources/META-INF/db/schema-41510to41600.sql`. But someone with more experience in this maybe should tell

blueorangutan commented:
Packaging result: :heavy_check_mark: centos7 :heavy_multiplication_x: centos8 :heavy_check_mark: debian. SL-JID 408

rp- commented:
@slavkap <3 That was it, thank you! Just weird that there is nowhere any SQL exception or anything similar...

slavkap commented:
That's great @rp- :) I think that there is no exception because of the inconsistency in DB with the renamed columns, and the SQL schema is not executed because you don't do an upgrade

rp- commented:
@rhtyd This is now ready for testing/review, except for one create path, because I don't know how to trigger it.
And that is the `createAsync` with a volume type of `TEMPLATE`. If you could guide me how I could trigger that (or maybe it isn't even necessary with our set capabilities).

The rest of the functionality should work as long as no LVM-Thick storagepool is used by linstor, as die plugin now requires linstor storage pools that support snapshots.
Also the included (copied solidfire) integration-test works on my setup.

rp- commented:
@rhtyd Pinging again, is there anything you still need from for this PR? Or what steps are missing to proceed?

shwstppr commented:
@blueorangutan package

blueorangutan commented:
@shwstppr a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

rhtyd commented:
Asking our RMs - @nvazquez @sureshanaparti to help kick tests and review.

@rp- We don't have a Linstor setup to test your changes so we can help kick regression smoketests and maybe you can provide some test results.

For example, you can run and share results of your smoketests in your env (or using virtualised https://github.com/shapeblue/mbx) with Linstor storage with KVM as the only primary storage. Or a set of manual tests. We can merge based on your provided test results and regression smoketests, and after all review items are addressed.

blueorangutan commented:
Packaging result: :heavy_check_mark: el7 :heavy_check_mark: el8 :heavy_check_mark: debian. SL-JID 537

rhtyd commented:
@blueorangutan test

blueorangutan commented:
@rhtyd a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian test result (tid-1265)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 55065 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr4994-t1265-kvm-centos7.zip
Intermittent failure detected: /marvin/tests/smoke/test_kubernetes_clusters.py
Intermittent failure detected: /marvin/tests/smoke/test_network.py
Intermittent failure detected: /marvin/tests/smoke/test_privategw_acl.py
Intermittent failure detected: /marvin/tests/smoke/test_routers_network_ops.py
Intermittent failure detected: /marvin/tests/smoke/test_vpc_redundant.py
Smoke tests completed. 85 look OK, 3 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_01_invalid_upgrade_kubernetes_cluster | `Failure` | 3615.27 | test_kubernetes_clusters.py
test_02_deploy_and_upgrade_kubernetes_cluster | `Failure` | 3615.37 | test_kubernetes_clusters.py
test_03_deploy_and_scale_kubernetes_cluster | `Failure` | 0.05 | test_kubernetes_clusters.py
test_04_basic_lifecycle_kubernetes_cluster | `Failure` | 0.04 | test_kubernetes_clusters.py
test_05_delete_kubernetes_cluster | `Failure` | 0.04 | test_kubernetes_clusters.py
test_07_deploy_kubernetes_ha_cluster | `Failure` | 0.04 | test_kubernetes_clusters.py
test_08_deploy_and_upgrade_kubernetes_ha_cluster | `Failure` | 0.04 | test_kubernetes_clusters.py
test_09_delete_kubernetes_ha_cluster | `Failure` | 0.04 | test_kubernetes_clusters.py
ContextSuite context=TestKubernetesCluster>:teardown | `Error` | 103.46 | test_kubernetes_clusters.py
test_02_RVR_Network_FW_PF_SSH_default_routes_egress_false | `Failure` | 339.72 | test_routers_network_ops.py
test_01_create_redundant_VPC_2tiers_4VMs_4IPs_4PF_ACL | `Failure` | 535.35 | test_vpc_redundant.py
test_03_create_redundant_VPC_1tier_2VMs_2IPs_2PF_ACL_reboot_routers | `Failure` | 463.85 | test_vpc_redundant.py

rp- commented:
@rhtyd 
> For example, you can run and share results of your smoketests in your env (or using virtualised https://github.com/shapeblue/mbx) with Linstor storage with KVM as the only primary storage. Or a set of manual tests. We can merge based on your provided test results and regression smoketests, and after all review items are addressed.

So I'm in the middle of setup a new cluster with just Linstor as primary storage, and in the setup wizard I noticed that only the options for (nfs, SharedMountPoint, RBD, CLVM, Gluster). So why are here solidfire, datera and others missing and is it ok to add the default nfs primary storage and afteward deleted it and add just a Linstor one? Or should I be able to already pick Linstor here?

rhtyd commented:
@rp- I would try and add Linstor in the list (API/UI) so you don't need to do the hack. As a user with your feature I should we able to simply select and use Linstor.

rhtyd commented:
@blueorangutan test

blueorangutan commented:
@rhtyd a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

sureshanaparti commented:
> @rhtyd
> 
> > For example, you can run and share results of your smoketests in your env (or using virtualised https://github.com/shapeblue/mbx) with Linstor storage with KVM as the only primary storage. Or a set of manual tests. We can merge based on your provided test results and regression smoketests, and after all review items are addressed.
> 
> So I'm in the middle of setup a new cluster with just Linstor as primary storage, and in the setup wizard I noticed that only the options for (nfs, SharedMountPoint, RBD, CLVM, Gluster). So why are here solidfire, datera and others missing and is it ok to add the default nfs primary storage and afteward deleted it and add just a Linstor one? Or should I be able to already pick Linstor here?

@rp- The protocol while adding primary storage from the UI, is basically used to build the url of the storage pool.

Protocol:
<img width="388" alt="Protocol_AddPrimaryStorage" src="https://user-images.githubusercontent.com/12028987/125801847-47e25bb2-bf7f-4272-b5e3-e589b40f899e.png">

Provider:
<img width="389" alt="Provider_AddPrimaryStorage" src="https://user-images.githubusercontent.com/12028987/125801880-ecfb8d40-e619-4257-b1d1-4dbc532efcee.png">

In your case, you can either use a new protocol 'Linstor', Otherwise use 'custom' protocol and make sure the 'Linstor' provider is selected, to general the url for the storage pool and pass to createStoragePool API.

sureshanaparti commented:
Hi @rp- can you please fix the conflicts. thanks.

rp- commented:
Short update: So I added Linstor protocol to the wizard and was kinda able to get the cluster running, but this time I used ubuntu and ZFS as backend storage. Which sadly surfaced some problems with ZFS's parent/child relation problematic on volumes and snapshots.
So after some thinking about this, we (Linstor) finally need a dedicated Clone/Copy resource API, which will handle the best possible clone solution for the used storage backend. So I have to first get this implemented into Linstor before I can continue on the cloudstack plugin.

I can meanwhile fix the conflict, but I guess you want to check/test the version after I finished the above mentioned API.

sureshanaparti commented:
> @rhtyd
> 
> > For example, you can run and share results of your smoketests in your env (or using virtualised https://github.com/shapeblue/mbx) with Linstor storage with KVM as the only primary storage. Or a set of manual tests. We can merge based on your provided test results and regression smoketests, and after all review items are addressed.
> 
> So I'm in the middle of setup a new cluster with just Linstor as primary storage, and in the setup wizard I noticed that only the options for (nfs, SharedMountPoint, RBD, CLVM, Gluster). So why are here solidfire, datera and others missing and is it ok to add the default nfs primary storage and afteward deleted it and add just a Linstor one? Or should I be able to already pick Linstor here?

@rp- missed to reply for the second part. In case you have any issues with changes to setup wizard (zone creation wizard) for Linstor, you can first add NFS as Primary Storage (as you said) from there, later add Linstor and remove NFS.

blueorangutan commented:
<b>Trillian test result (tid-1275)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 43680 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr4994-t1275-kvm-centos7.zip
Intermittent failure detected: /marvin/tests/smoke/test_deploy_virtio_scsi_vm.py
Intermittent failure detected: /marvin/tests/smoke/test_privategw_acl.py
Intermittent failure detected: /marvin/tests/smoke/test_routers_network_ops.py
Intermittent failure detected: /marvin/tests/smoke/test_vpc_redundant.py
Smoke tests completed. 86 look OK, 2 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_03_vpc_privategw_restart_vpc_cleanup | `Failure` | 399.55 | test_privategw_acl.py
test_01_create_redundant_VPC_2tiers_4VMs_4IPs_4PF_ACL | `Failure` | 961.85 | test_vpc_redundant.py
test_03_create_redundant_VPC_1tier_2VMs_2IPs_2PF_ACL_reboot_routers | `Failure` | 476.06 | test_vpc_redundant.py

rp- commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/storage/LinstorStorageAdaptor.java:
```diff
@@ -0,0 +1,578 @@
+        try {
+            final String rscName = getLinstorRscName(name);
+            s_logger.debug("Linstor: delete resource definition " + rscName);
+            ApiCallRcList answers = api.resourceDefinitionDelete(rscName);
```
@slavkap No, you would need to delete the snapshots before you could delete the volume.

andrijapanicsb commented:
@rp- please solve the conflict 
thanks

andrijapanicsb commented:
@rp- being the lazy bastard I am (haven't read the design cwiki), do you have a list of ACS features what is possible (and what not) with this current/initial implementation? To have it outlined somewhere

rhtyd commented:
@rp- can you fix the conflicts? Since smoketest round have passed, if you can provide details of your tests - we can do another round of review and regression smoketests. We can merge it if we've 2xreview lgtms and smoketests pass.

rp- commented:
@andrijapanicsb @rhtyd 
As stated above, I noticed while testing on a new cluster with ZFS, that current implemention doesn't work out as planned.
So I'm currently implementing a new feature in Linstor(Clone) to fix this and I'm nearly done with it. Have to do some more testing and writing tests on Linstor side.
After this is done, I'll update the plugin and do a force push of the plugin driver. It probably doesn't make sense to review it in the current state.
I'll try to get the code updated next week.

nvazquez commented:
Thanks for the update @rp-

DaanHoogland commented:
@rp- I'm putting this back into draft because of your last remark.

rp- commented:
@andrijapanicsb @rhtyd

So I was able to finish the "Clone" feature in Linstor and continued to integrate this new functionality into the Linstor volume plugin. But while testing I encountered the following problem and I'm not sure who has to handle it(or how):

I tried to setup a cluster with just the Linstor primary storage and my problem is now the format of the system-vm image.
For kvm it is in the qcow2 format and as other storage providers Linstor will report RAW as supported and used format.
But what I end up with is a qcow2 file stored within our block device(which isn't really that much of a problem) but libvirt will see the disk as RAW and will of course fail to boot the qcow2 image.

So either the image gets converted while creating the first(hidden) base volume or I would need to tell libvirt somewhere that this is indeed a qcow2 image.

Do other storage plugins handle this correctly? Like RDB, Powerflex...? As I saw in the code they also only handle RAW format.

rhtyd commented:
@rp- yes, other plugins for example in case of ScaleIO and Ceph, the template is converted to RAW; the specific disk/file/drive is mounted on KVM host on which qemu-img is used to convert a guest template from a valid source (qcow2, raw...) to raw.

/cc @sureshanaparti  (looping in Suresh who wrote the ScaleIO storage plugin)

rhtyd commented:
Ping @rp- any update? The 4.16 freeze may be around the corner soon (say 2 weeks)

rp- commented:
yes sorry, this clone feature in Linstor took a bit longer than expected, but a first version is now released in Linstor 1.15.0-rc1.
I also updated the plugin for the new API and did some tests and think it is ready for review now.

rhtyd commented:
Thanks for updating @rp- I'll kick smoketests and happy to merge this if the PR passes smoketests. Can you share the tests you've performed with your plugin, since we don't have your env/setup or a Linstor setup we'll rely on you for functional testing. cc @andrijapanicsb 

@blueorangutan package

blueorangutan commented:
@rhtyd a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: :heavy_check_mark: el7 :heavy_check_mark: el8 :heavy_check_mark: debian :heavy_check_mark: suse15. SL-JID 1078

rp- commented:
I have ran the test from `test/integration/plugins/linstor/test_linstor_volumes.py` and it succeeded and did some manual setup with a linstor only storage pool and creating a VM in there(+ the system VM's).
I'll try to get additional testing done in the next week.

rhtyd commented:
@blueorangutan test

blueorangutan commented:
@rhtyd a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian Build Failed (tid-1858)<b/>

rhtyd commented:
@blueorangutan test

blueorangutan commented:
@rhtyd a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian Build Failed (tid-1863)<b/>

rhtyd commented:
@blueorangutan test

blueorangutan commented:
@rhtyd a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

rhtyd commented:
@blueorangutan test

blueorangutan commented:
@rhtyd a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian test result (tid-1934)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 36117 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr4994-t1934-kvm-centos7.zip
Smoke tests completed. 89 look OK, 0 have errors
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---

rhtyd commented:
Since smoketests have passed, may I request one of the reviewer to help with a final review - maybe @sureshanaparti and @slavkap  since you've more experience around storage?
@rp- we can merge this now as soon as we've 2xreviews/lgtm; any additional tests/results from you would also be great.

slavkap commented on plugins/storage/volume/linstor/src/main/java/org/apache/cloudstack/storage/datastore/driver/LinstorPrimaryDataStoreDriverImpl.java:
```diff
@@ -0,0 +1,720 @@
+    }
+
+    @Override
+    public void resize(DataObject data, AsyncCompletionCallback<CreateCmdResult> callback)
```
@rp- I'm not sure if I'm right, but when you do a live resize, you're resizing the volume only on Linstor, but the guest/hypervisor is not informed. That's because the Linstor primary storage is not managed, and you have to notify qemu from here. Please excuse me if I'm wrong

rp- replied:
I tried resizing, indeed online resizing isn't noticed by the guest. But if i power down and start the guest again. it has the new size available. I looked into other storage drivers, but didn't really see were they would notify the virtualization platform of the size change?

andrijapanicsb replied:
Do we do this for other storage types? I'm not aware we support online disk resize for other storage platforms, but I may be wrong.

slavkap replied:
@rp-, most of the plugins in CloudStack are for managed storage. In your case I think it will be useful for you to check the default driver implementation - `CloudStackPrimaryDataStoreDriverImpl.resize()` and maybe you have to modify `LibvirtResizeVolumeCommandWrapper` or to add a custom `CommandWrapper`.

slavkap replied:
@andrijapanicsb, we at StorPool support live resize. It is supported for NFS/Ceph also

rp- replied:
I added the notify call, but it didn't make a difference for my alpine guest.

slavkap replied:
@rp- is there any output when 
`/usr/share/cloudstack-common/scripts/storage/qcow2/resizevolume.sh -s
 newSize -c currentSize -p volumePath -t NOTIFYONLY -r shrink -v vmName` is executed?

rp- replied:
just got the following:
```
Sep 08 08:52:51 cloudstack-comp2.linbit java[6215]: INFO  [resource.wrapper.LibvirtResizeVolumeCommandWrapper] (agentRequest-Handler-1:) (logid:08a8fb35) Invoking resize script to handle type NOTIFYONLY
Sep 08 08:52:51 cloudstack-comp2.linbit java[6215]: DEBUG [resource.wrapper.LibvirtResizeVolumeCommandWrapper] (agentRequest-Handler-1:) (logid:08a8fb35) Executing: /usr/share/cloudstack-common/scripts/storage
/qcow2/resizevolume.sh -s 10737418240 -c 9663676416 -p /dev/drbd1001 -t NOTIFYONLY -r false -v i-2-253-VM
Sep 08 08:52:51 cloudstack-comp2.linbit java[6215]: DEBUG [resource.wrapper.LibvirtResizeVolumeCommandWrapper] (agentRequest-Handler-1:) (logid:08a8fb35) Executing while with timeout : 7200000
Sep 08 08:52:51 cloudstack-comp2.linbit java[6215]: DEBUG [resource.wrapper.LibvirtResizeVolumeCommandWrapper] (agentRequest-Handler-1:) (logid:08a8fb35) Execution is successful.
```

slavkap replied:
Hmm, this tells that the notification of the guest was successful. Is there any change when you run:
`virsh domblkinfo --device /dev/drbd1001 --domain i-2-253-VM --human`
or
`lsblk` from the guest?

rp- replied:
```
[root@cloudstack-comp2 lib]# virsh domblkinfo --device /dev/drbd1001
 --domain i-2-253-VM --human
Capacity:       10.002 GiB
Allocation:     1.102 GiB
Physical:       10.002 GiB
```

lsblk in guest still tells me the old size

rhtyd commented:
@blueorangutan package

blueorangutan commented:
@rhtyd a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: :heavy_multiplication_x: el7 :heavy_multiplication_x: el8 :heavy_multiplication_x: debian :heavy_multiplication_x: suse15. SL-JID 1146

nvazquez commented:
@blueorangutan package

blueorangutan commented:
@nvazquez a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: :heavy_multiplication_x: el7 :heavy_multiplication_x: el8 :heavy_multiplication_x: debian :heavy_multiplication_x: suse15. SL-JID 1149

nvazquez commented:
Hi @rp- seems there is an issue on the packaging, can you please check?

rp- commented:
are there any error logs? or how could I try to package myself?
maybe a rebase on `main` is enough?

rp- commented:
@nvazquez I did rebase against current main, maybe retry the builds, otherwise I would need some error logs

nvazquez commented:
@blueorangutan package

blueorangutan commented:
@nvazquez a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: :heavy_multiplication_x: el7 :heavy_multiplication_x: el8 :heavy_multiplication_x: debian :heavy_multiplication_x: suse15. SL-JID 1200

nvazquez commented:
@rp- there is a failure on a unit test, there may be other failures as well but were not executed. Please check if you are building using the `-DskipTests` flag, you may want to remove it in order to run the unit tests

````
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running com.cloud.hypervisor.kvm.resource.LibvirtDomainXMLParserTest
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.237 s - in com.cloud.hypervisor.kvm.resource.LibvirtDomainXMLParserTest
[INFO] Running com.cloud.hypervisor.kvm.resource.LibvirtComputingResourceTest
log4j:WARN No appenders could be found for logger (com.cloud.agent.properties.AgentPropertiesFileHandler).
log4j:WARN Please initialize the log4j system properly.
log4j:WARN See http://logging.apache.org/log4j/1.2/faq.html#noconfig for more info.
[ERROR] Tests run: 187, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 6.037 s <<< FAILURE! - in com.cloud.hypervisor.kvm.resource.LibvirtComputingResourceTest
[ERROR] testResizeVolumeCommand(com.cloud.hypervisor.kvm.resource.LibvirtComputingResourceTest)  Time elapsed: 0.027 s  <<< FAILURE!
java.lang.AssertionError
	at com.cloud.hypervisor.kvm.resource.LibvirtComputingResourceTest.testResizeVolumeCommand(LibvirtComputingResourceTest.java:4871)
````

rp- commented:
oh sorry, I forgot about unittests.
Fixed, force pushed and also added a unittest for linstor there.

nvazquez commented:
Thanks @rp- 
@blueorangutan package

blueorangutan commented:
@nvazquez a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: :heavy_check_mark: el7 :heavy_check_mark: el8 :heavy_check_mark: debian :heavy_check_mark: suse15. SL-JID 1218

nvazquez commented:
@blueorangutan test

blueorangutan commented:
@nvazquez a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian test result (tid-2019)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 36596 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr4994-t2019-kvm-centos7.zip
Smoke tests completed. 89 look OK, 0 have errors
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---

rhtyd commented:
@rp- can you fix the conflicts? I think this is otherwise good to merge.

rp- commented:
> @rp- can you fix the conflicts? I think this is otherwise good to merge.

done

blueorangutan commented:
@rhtyd a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: :heavy_check_mark: el7 :heavy_check_mark: el8 :heavy_check_mark: debian :heavy_check_mark: suse15. SL-JID 1242

nvazquez commented:
@blueorangutan test

blueorangutan commented:
@nvazquez a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian Build Failed (tid-2048)<b/>

nvazquez commented:
@blueorangutan test

blueorangutan commented:
@nvazquez a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

rhtyd commented:
If smoketests pass, I don't mind merging this. As we don't have the storage setup, we'll rely on @rp- for providing tests.

rhtyd commented on plugins/storage/volume/linstor/src/main/java/org/apache/cloudstack/storage/datastore/util/LinstorUtil.java:
```diff
@@ -0,0 +1,65 @@
+package org.apache.cloudstack.storage.datastore.util;
```
@rp- general remark - you want to add the Apache License headers to all new files.

rhtyd reviewed (approved):
LGTM, some changes needed @rp-:
- Add Apache License 2.0 headers on all new files
- Can you check if changes in the cloudbyte storage plugin necessary
- Let me also ask my colleague @sureshanaparti to review (he has experience in writing storage plugins)

rhtyd commented:
@blueorangutan test centos7 vmware-67u3

rhtyd commented:
@blueorangutan test centos7 xenserver-71

blueorangutan commented:
@rhtyd a Trillian-Jenkins test job (centos7 mgmt + xenserver-71) has been kicked to run smoke tests

rhtyd commented:
@blueorangutan test centos7 vmware-67u3

blueorangutan commented:
@rhtyd a Trillian-Jenkins test job (centos7 mgmt + vmware-67u3) has been kicked to run smoke tests

rp- commented:
@rhtyd 

> LGTM, some changes needed @rp-:
> * Add Apache License 2.0 headers on all new files

Fixed/Added

> * Can you check if changes in the cloudbyte storage plugin necessary

Yes, unfortunately the cloudbyte plugin uses an old version of jersey client and I introduced a newer one which didn't build
correctly if mixed. As far as I remember.
I also assume it is in common interest to only have one version of jersey as dependency.

rhtyd commented:
Thanks @rp-

blueorangutan commented:
<b>Trillian test result (tid-2047)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 34981 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr4994-t2047-kvm-centos7.zip
Smoke tests completed. 89 look OK, 0 have errors
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---

nvazquez commented:
Thanks @rp- @rhtyd - smoke tests passed, let's wait for @sureshanaparti's review

rhtyd commented:
@blueorangutan package

blueorangutan commented:
@rhtyd a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
<b>Trillian test result (tid-2049)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 37354 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr4994-t2049-kvm-centos7.zip
Smoke tests completed. 89 look OK, 0 have errors
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---

blueorangutan commented:
Packaging result: :heavy_check_mark: el7 :heavy_check_mark: el8 :heavy_check_mark: debian :heavy_check_mark: suse15. SL-JID 1273

rhtyd commented:
@blueorangutan test

blueorangutan commented:
@rhtyd a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian test result (tid-2058)</b>
Environment: vmware-67u3 (x2), Advanced Networking with Mgmt server 7
Total time taken: 41973 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr4994-t2058-vmware-67u3.zip
Smoke tests completed. 89 look OK, 0 have errors
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---

rhtyd commented:
Merged based on 2xlgtms, and on smoketests and Travis passing. We haven't tested Linstor itself which is tested by @rp-

sureshanaparti commented:
> Thanks @rp- @rhtyd - smoke tests passed, let's wait for @sureshanaparti's review

@nvazquez reviewed the PR, LGTM.

rhtyd commented:
@rp- @Philipp-Reisner can you check a potential Linstor regression - https://markmail.org/message/tc4qkm2izzcjqahn 
I'll tag/cc you in the email thread too.

soptecsn commented:
We are trying to deploy linstor driver for primary storage and it is not wokring. We have configured 3 linstor nodes with ZFS, ACS 4.16

First we get an error that linstor was not installed in the host, so we install it and the error disappear, we think is not correct that Host has to be installed with the software because they just to be a "client". But there aren't any documentation about this.

Then, when that error disappear, we try to create a VM and the log in the host show this:

INFO
 [kvm.storage.LinstorStorageAdaptor] (agentRequest-Handler-1:null)
 (logid:1fb3773c)
 Linstor: Created drbd device: /dev/zvol/Linstor-ST04/cs-c3bbbcbb-142a-4f28-bf1b-fd0748c7a6f7_00000
 ERROR [kvm.storage.LinstorStorageAdaptor] (agentRequest-Handler-1:null)
 (logid:1fb3773c)
 org.apache.cloudstack.utils.qemu.QemuImgException: qemu-img:
 /dev/zvol/Linstor-ST04/cs-c3bbbcbb-142a-4f28-bf1b-fd0748c7a6f7_00000:
 error while
 converting raw:
 Could not create file:
 No such file
or
 directory

When we go to the linstor dashboard, we see that Linstor creates a Resource Definition, then a resource into and finally the volumen, until this moment it all looking good, but inmediatly, deletes all resources created (resource definition, resource and volume).

Linstor and Cloudstack are connecting fine when we add the primary storage.

Error from the linstor side:

 
        ERROR REPORT 617B1236-00000-003014

============================================================

Application:                        LINBIT® LINSTOR
Module:                             Controller
Version:                            1.15.0
Build ID:                           7a46abd8becd8027357d06f5990a272224a71e81
Build time:                         2021-09-23T06:21:28+00:00
Error time:                         2021-11-01 20:59:15
Node:                               linstor-st01
Peer:                               RestClient(10.0.10.26; 'Swagger-Codegen/1.0.0/java')

============================================================

Reported error:
===============

Category:                           RuntimeException
Class name:                         ApiRcException
Class canonical name:               com.linbit.linstor.core.apicallhandler.response.ApiRcException
Generated at:                       Method 'loadRscDfn', Source file 'CtrlApiDataLoader.java', Line #188

Error message:                      Resource definition 'cs-c3bbbcbb-142a-4f28-bf1b-fd0748c7a6f7' not found.

Error context:
    Resource definition 'cs-c3bbbcbb-142a-4f28-bf1b-fd0748c7a6f7' not found.

Asynchronous stage backtrace:

    Error has been observed at the following site(s):
    	|_ checkpoint ⇢ Clone resource-definition
    Stack trace:

Call backtrace:

    Method                                   Native Class:Line number
    loadRscDfn                               N      com.linbit.linstor.core.apicallhandler.controller.CtrlApiDataLoader:188

Suppressed exception 1 of 1:
===============
Category:                           RuntimeException
Class name:                         OnAssemblyException
Class canonical name:               reactor.core.publisher.FluxOnAssembly.OnAssemblyException
Generated at:                       Method 'loadRscDfn', Source file 'CtrlApiDataLoader.java', Line #188

Error message:                      
Error has been observed at the following site(s):
	|_ checkpoint ⇢ Clone resource-definition
Stack trace:

Error context:
    Resource definition 'cs-c3bbbcbb-142a-4f28-bf1b-fd0748c7a6f7' not found.

Call backtrace:

    Method                                   Native Class:Line number
    loadRscDfn                               N      com.linbit.linstor.core.apicallhandler.controller.CtrlApiDataLoader:188
    loadRscDfn                               N      com.linbit.linstor.core.apicallhandler.controller.CtrlApiDataLoader:162
    cloneRscDfnInTransaction                 N      com.linbit.linstor.core.apicallhandler.controller.CtrlRscDfnApiCallHandler:850
    lambda$cloneRscDfn$6                     N      com.linbit.linstor.core.apicallhandler.controller.CtrlRscDfnApiCallHandler:723
    doInScope                                N      com.linbit.linstor.core.apicallhandler.ScopeRunner:147
    lambda$fluxInScope$0                     N      com.linbit.linstor.core.apicallhandler.ScopeRunner:75
    call                                     N      reactor.core.publisher.MonoCallable:91
    trySubscribeScalarMap                    N      reactor.core.publisher.FluxFlatMap:126
    subscribeOrReturn                        N      reactor.core.publisher.MonoFlatMapMany:49
    subscribe                                N      reactor.core.publisher.Flux:8343
    onNext                                   N      reactor.core.publisher.MonoFlatMapMany$FlatMapManyMain:188
    request                                  N      reactor.core.publisher.Operators$ScalarSubscription:2344
    onSubscribe                              N      reactor.core.publisher.MonoFlatMapMany$FlatMapManyMain:134
    subscribe                                N      reactor.core.publisher.MonoCurrentContext:35
    subscribe                                N      reactor.core.publisher.Flux:8357
    onNext                                   N      reactor.core.publisher.MonoFlatMapMany$FlatMapManyMain:188
    request                                  N      reactor.core.publisher.Operators$ScalarSubscription:2344
    onSubscribe                              N      reactor.core.publisher.MonoFlatMapMany$FlatMapManyMain:134
    subscribe                                N      reactor.core.publisher.MonoCurrentContext:35
    subscribe                                N      reactor.core.publisher.Flux:8357
    onNext                                   N      reactor.core.publisher.MonoFlatMapMany$FlatMapManyMain:188
    onNext                                   N      reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber:121
    complete                                 N      reactor.core.publisher.Operators$MonoSubscriber:1782
    onComplete                               N      reactor.core.publisher.MonoCollect$CollectSubscriber:152
    onComplete                               N      reactor.core.publisher.MonoFlatMapMany$FlatMapManyInner:252
    checkTerminated                          N      reactor.core.publisher.FluxFlatMap$FlatMapMain:838
    drainLoop                                N      reactor.core.publisher.FluxFlatMap$FlatMapMain:600
    drain                                    N      reactor.core.publisher.FluxFlatMap$FlatMapMain:580
    onComplete                               N      reactor.core.publisher.FluxFlatMap$FlatMapMain:457
    checkTerminated                          N      reactor.core.publisher.FluxFlatMap$FlatMapMain:838
    drainLoop                                N      reactor.core.publisher.FluxFlatMap$FlatMapMain:600
    innerComplete                            N      reactor.core.publisher.FluxFlatMap$FlatMapMain:909
    onComplete                               N      reactor.core.publisher.FluxFlatMap$FlatMapInner:1013
    onComplete                               N      reactor.core.publisher.FluxMap$MapSubscriber:136
    onComplete                               N      reactor.core.publisher.Operators$MultiSubscriptionSubscriber:2016
    onComplete                               N      reactor.core.publisher.FluxSwitchIfEmpty$SwitchIfEmptySubscriber:78
    complete                                 N      reactor.core.publisher.FluxCreate$BaseSink:438
    drain                                    N      reactor.core.publisher.FluxCreate$BufferAsyncSink:784
    complete                                 N      reactor.core.publisher.FluxCreate$BufferAsyncSink:732
    drainLoop                                N      reactor.core.publisher.FluxCreate$SerializedSink:239
    drain                                    N      reactor.core.publisher.FluxCreate$SerializedSink:205
    complete                                 N      reactor.core.publisher.FluxCreate$SerializedSink:196
    apiCallComplete                          N      com.linbit.linstor.netcom.TcpConnectorPeer:455
    handleComplete                           N      com.linbit.linstor.proto.CommonMessageProcessor:363
    handleDataMessage                        N      com.linbit.linstor.proto.CommonMessageProcessor:287
    doProcessInOrderMessage                  N      com.linbit.linstor.proto.CommonMessageProcessor:235
    lambda$doProcessMessage$3                N      com.linbit.linstor.proto.CommonMessageProcessor:220
    subscribe                                N      reactor.core.publisher.FluxDefer:46
    subscribe                                N      reactor.core.publisher.Flux:8357
    onNext                                   N      reactor.core.publisher.FluxFlatMap$FlatMapMain:418
    drainAsync                               N      reactor.core.publisher.FluxFlattenIterable$FlattenIterableSubscriber:414
    drain                                    N      reactor.core.publisher.FluxFlattenIterable$FlattenIterableSubscriber:679
    onNext                                   N      reactor.core.publisher.FluxFlattenIterable$FlattenIterableSubscriber:243
    drainFused                               N      reactor.core.publisher.UnicastProcessor:286
    drain                                    N      reactor.core.publisher.UnicastProcessor:329
    onNext                                   N      reactor.core.publisher.UnicastProcessor:408
    next                                     N      reactor.core.publisher.FluxCreate$IgnoreSink:618
    drainLoop                                N      reactor.core.publisher.FluxCreate$SerializedSink:248
    next                                     N      reactor.core.publisher.FluxCreate$SerializedSink:168
    processInOrder                           N      com.linbit.linstor.netcom.TcpConnectorPeer:373
    doProcessMessage                         N      com.linbit.linstor.proto.CommonMessageProcessor:218
    lambda$processMessage$2                  N      com.linbit.linstor.proto.CommonMessageProcessor:164
    onNext                                   N      reactor.core.publisher.FluxPeek$PeekSubscriber:177
    runAsync                                 N      reactor.core.publisher.FluxPublishOn$PublishOnSubscriber:439
    run                                      N      reactor.core.publisher.FluxPublishOn$PublishOnSubscriber:526
    call                                     N      reactor.core.scheduler.WorkerTask:84
    call                                     N      reactor.core.scheduler.WorkerTask:37
    run                                      N      java.util.concurrent.FutureTask:264
    run                                      N      java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask:304
    runWorker                                N      java.util.concurrent.ThreadPoolExecutor:1128
    run                                      N      java.util.concurrent.ThreadPoolExecutor$Worker:628
    run                                      N      java.lang.Thread:829


END OF ERROR REPORT.

nvazquez commented:
Hi @rp- @Philipp-Reisner can you please advise on the reported issue? Could be a configuration issue or a bug in CloudStack that needs fixing?

rp- commented:
@nvazquez I'm on it via a different communication channel

But I suspect a configuration problem, I'll report back here what solved the case.
